detox/runners/jest/DetoxAdapterImpl.js

```diff
@@ -22,8 +22,10 @@ class DetoxAdapterImpl {
       throw new DetoxRuntimeError(this._describeInitError());
     }
 
+    const currentTest = this._currentTest;
+
     await this._flush();
-    await this.detox.beforeEach(this._currentTest);
+    await this.detox.beforeEach(currentTest);
```

Comment:
So... sometimes `this._currentTest` is cleared waiting for the (resolved!) promise returned by `_flush()`?!

Reply:
Seems like your hunch is very close to the truth. There are multiple cursors stuck between `await this._flush();` and `await this.detox.beforeEach(currentTest);`, due to stuck `detox.afterEach()` async calls. After `detox.cleanup()` they unstick and we have weird `this.detox.beforeEach()` calls after all the battle has been already lost. 🤦 

Actually that makes me think that a bigger fix would be to make calls to detox lifecycle methods blocking, either unless they execute or `detox.cleanup()` is called.